pyfr/integrators/dual/base.py

```diff
@@ -3,24 +3,116 @@
 from abc import abstractmethod
 
 from pyfr.integrators.base import BaseIntegrator
+from pyfr.util import memoize, proxylist
 
 
 class BaseDualIntegrator(BaseIntegrator):
     formulation = 'dual'
 
-    def __init__(self, *args, **kwargs):
-        super().__init__(*args, **kwargs)
+    def __init__(self, backend, systemcls, rallocs, mesh, initsoln, cfg):
+        super().__init__(backend, systemcls, rallocs, mesh, initsoln, cfg)
+        sect = 'solver-time-integrator'
 
-        self._dtau = self.cfg.getfloat('solver-time-integrator', 'pseudo-dt')
+        self._dtau = self.cfg.getfloat(sect, 'pseudo-dt')
         self.dtaumin = 1.0e-12
 
+        # Dual Integrator treats following variables as lists to accommodate MG
+        self.dualsystems = []
+        self._dualregs = []
+        self._dualregidx = []
+        self._dualidxcurr = []
+        self._dualidxprev = []
+
+        self.levels = 1
+        nreg = self._stepper_nregs
+
+        # Check whether to employ multigrid
+        self.multigrid = self.cfg.get(sect, 'controller') == 'multigrid'
+
+        if self.multigrid:
+            # Multigrid requires 3 additional registers
+            nreg += 3
+            # Number of iterations at each stage
+            self._leveliters = self.cfg.getintlist(sect, 'mgcycle')
+            # Number of levels
+            self.levels = len(self._leveliters)
+
+        for level in range(self.levels):
+            self.dualsystems.append(systemcls(backend, rallocs, mesh,
+                                              initsoln, nreg, cfg, level))
+            self._dualregs.append(self._get_reg_banks(nreg, level)[0])
+            self._dualregidx.append(self._get_reg_banks(nreg, level)[1])
+            self._dualidxcurr.append(0)
```

Comment:
Outside of the loop you can just do:

self._dualidxcurr = [0]*self.levels
self._dualidxprev = [0]*self.levels